src/Engine/Component/ParticleComponents/ParticleEffect.cpp

```diff
@@ -0,0 +1,393 @@
+        u = normalize(cross(directionNorm, randVec));
+    }
+    return u;
+}
+
+
+float ParticleEffect::getRandom(float low, float high) {
+    return low + static_cast <float> (rand()) / (static_cast <float> (RAND_MAX / (high - low)));
+}
+
+Vector<ParticleEffect::Particle*> ParticleEffect::generateParticles() {
+    Vector<Particle*> vp = Vector<Particle*>();
+    for (int i = 0; i < m_effectParams->n; i++) {
+        Particle *p = makeParticle(i);
+        vp.push_back(p);
+    }
+    return vp;
+}
+
+// This and the inits need to eventually be cleaned up
+ParticleEffect::Particle* ParticleEffect::makeParticle(int i) {
+    Particle *p = new Particle();
+    
+    initParticle(p, i, 0, 0);
+
+    //Determine if active/inactive
+    if (m_effectParams->rate == 0.0f) {
+        p->active = true;
+    }
+    else {
+        p->active = false;
+    }
+
+    // Get Orientation ID
+    if (m_effectParams->orientations == 0) {
+        p->orientationID = 0;
+    }
+    else {
+        p->orientationID = p->i % m_effectParams->orientations;
+    }
+    
+
+    initPosition(p);
+
+    initVelocity(p);
+
+    return p;
+}
+
+void ParticleEffect::initParticle(Particle *p, int i, int meshID, int modelTextureID) {
+    p->i = i;
+    p->meshID = meshID;
+    p->modelTextureID = modelTextureID;
+    p->life = 0.0f;
+}
+
+void ParticleEffect::initVelocity(Particle *p) {
+    switch (m_effectParams->type) {
+        case ParticleEffect::SPHERE : {
+            sphereMotion(p);
+            break;
+        }
+        case ParticleEffect::DISK : {
+            diskMotion(p);
+            break;
+        }
+        case ParticleEffect::CONE: {
+            coneMotion(p);
+            break;
+        }
+    }
+
+    //Add random variance if specified
+    if (m_effectParams->variance != 0.0f) {
+        glm::vec3 randVec = m_effectParams->variance * normalize(glm::ballRand(1.0f));
+        p->velocity += randVec;
+    }
+
+    if (m_effectParams->randomDistribution) {
+        p->velocity *= getRandom(0.25f, 1.0f);
+    }
+
+    //set speed
+    p->velocity *= m_effectParams->magnitude;
+}
+
+void ParticleEffect::initPosition(Particle *p) {
+    p->position = m_anchor + m_offset + m_life * m_velocity;
+}
+
+Vector<glm::vec3> * ParticleEffect::getActiveParticlePositions() {
+    Vector<glm::vec3> *positions = new Vector<glm::vec3>();
+    for (int i = 0; i < m_effectParams->n; i++){
+        if (m_particles[i]->active) {
+            positions->push_back(m_particles[i]->position);
+        }
+    }
+    return positions;
+}
+
+Vector<int> * ParticleEffect::getActiveParticleOrientationIDs() {
+    Vector<int> *orientationIDs = new Vector<int>();
+    for (int i = 0; i < m_effectParams->n; i++) {
+        if (m_particles[i]->active) {
+            orientationIDs->push_back(m_particles[i]->orientationID);
+        }
+    }
+    return orientationIDs;
+}
+
+Vector<int> ParticleEffect::getActiveMap() {
+    Vector<int> map = Vector<int>();
```

Comment:
I'm not a fan of returning vectors by value. I think this should either be a new Vector or it should return some sort of reference.

Reply:
this is actually fine so long as when you return you do std::move

Reply:
also, this can (should) just be Vector\<int\> map;

Reply:
or you can pass the "return" vector in as a non-const reference and populate it. This is C-style but works fine